### ISSUE TITLE
Build farm fails to copy a file, gives Java backtrace instead

### ISSUE THREAD

**Issue by mjcarroll** [ros2/build_farmer#101]
Build farm fails to copy a file, gives Java backtrace instead.  This happens when trying to look at GCC compiler warning output for the rclpy job.

Job: https://ci.ros2.org/job/ci_linux/4137/
Warnings page: https://ci.ros2.org/job/ci_linux/4137/warnings23Result/new/source.-227773852282226723/#237

```
01 Copying the source file '/home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c' from the workspace to the build folder '2f48a1c5.tmp' on the Jenkins master failed.
02 Is the file '/home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c' a valid filename?
03 If you are building on a slave: please check if the file is accessible under '$JENKINS_HOME/[job-name]//home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c'
04 If you are building on the master: please check if the file is accessible under '$JENKINS_HOME/[job-name]/workspace//home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c'
05 java.io.IOException: Failed to copy /home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c to /var/lib/jenkins/jobs/ci_linux/builds/4137/workspace-files/2f48a1c5.tmp
06   at hudson.FilePath.copyTo(FilePath.java:2002)
07   at hudson.plugins.analysis.util.Files.copyFilesWithAnnotationsToBuildFolder(Files.java:80)
08   at hudson.plugins.analysis.core.HealthAwareRecorder.copyFilesWithAnnotationsToBuildFolder(HealthAwareRecorder.java:351)
09   at hudson.plugins.analysis.core.HealthAwarePublisher.perform(HealthAwarePublisher.java:91)
10   at hudson.plugins.analysis.core.HealthAwareRecorder.perform(HealthAwareRecorder.java:298)
11   at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
12   at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
13   at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
14   at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
15   at hudson.model.Build$BuildExecution.post2(Build.java:186)
16   at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
17   at hudson.model.Run.execute(Run.java:1749)
18   at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
19   at hudson.model.ResourceController.execute(ResourceController.java:97)
20   at hudson.model.Executor.run(Executor.java:429)
21 Caused by: java.io.IOException: remote file operation failed: /home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c at hudson.remoting.Channel@1b9ba5a0:JNLP4-connect connection from ip-172-30-0-87.ec2.internal/172.30.0.87:58722: java.nio.file.NoSuchFileException: /home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c
22   at hudson.FilePath.act(FilePath.java:993)
23   at hudson.FilePath.act(FilePath.java:975)
24   at hudson.FilePath.copyTo(FilePath.java:2023)
25   at hudson.FilePath.copyTo(FilePath.java:1999)
26   ... 14 more
27 Caused by: java.nio.file.NoSuchFileException: /home/rosbuild/ci_scripts/ws/src/ros2/rclpy/rclpy/src/rclpy/_rclpy.c
28   at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
29   at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
30   at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
31   at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
32   at java.nio.file.Files.newByteChannel(Files.java:361)
33   at java.nio.file.Files.newByteChannel(Files.java:407)
34   at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
35   at java.nio.file.Files.newInputStream(Files.java:152)
36   at hudson.FilePath$41.invoke(FilePath.java:2026)
37   at hudson.FilePath$41.invoke(FilePath.java:2023)
38   at hudson.FilePath$FileCallableWrapper.call(FilePath.java:2760)
39   at hudson.remoting.UserRequest.perform(UserRequest.java:210)
40   at hudson.remoting.UserRequest.perform(UserRequest.java:53)
41   at hudson.remoting.Request$2.run(Request.java:358)
42   at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
43   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
44   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
45   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
46   at hudson.remoting.Engine$1$1.run(Engine.java:94)
47   at java.lang.Thread.run(Thread.java:748)
48   Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from ip-172-30-0-87.ec2.internal/172.30.0.87:58722
49     at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1693)
50     at hudson.remoting.UserResponse.retrieve(UserRequest.java:310)
51     at hudson.remoting.Channel.call(Channel.java:908)
52     at hudson.FilePath.act(FilePath.java:986)
53     at hudson.FilePath.act(FilePath.java:975)
54     at hudson.FilePath.copyTo(FilePath.java:2023)
55     at hudson.FilePath.copyTo(FilePath.java:1999)
56     at hudson.plugins.analysis.util.Files.copyFilesWithAnnotationsToBuildFolder(Files.java:80)
57     at hudson.plugins.analysis.core.HealthAwareRecorder.copyFilesWithAnnotationsToBuildFolder(HealthAwareRecorder.java:351)
58     at hudson.plugins.analysis.core.HealthAwarePublisher.perform(HealthAwarePublisher.java:91)
59     at hudson.plugins.analysis.core.HealthAwareRecorder.perform(HealthAwareRecorder.java:298)
60     at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
61     at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
62     at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
63     at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
64     at hudson.model.Build$BuildExecution.post2(Build.java:186)
65     at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
66     at hudson.model.Run.execute(Run.java:1749)
67     at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
68     at hudson.model.ResourceController.execute(ResourceController.java:97)
69     at hudson.model.Executor.run(Executor.java:429)
```

**Comment by nuclearsandwich** [comment]
It looks like the warning is extracting a filename from the log and assuming that the file is present on the build machine at that location. Because gcc is running within the container, the path the the file is absolute within that container rather than the host filesystem. I'm not sure if this is something that worked previously and I broke it with the migration to xenial or if this has never worked for contained linux builds.

Depending on how this warning plugin works we might be able to allow it to munge the paths to find files on the host filesystem.

**Comment by nuclearsandwich** [comment]
@mikaelarguedas or @dirk-thomas do you know if this particular ui resolved or was working before the migration. If it was, I'm not sure how.

**Comment by mikaelarguedas** [comment]
I don't think I ever used that ui feature so cannot say if it worked before the migration

**Comment by dirk-thomas** [comment]
For some files showing the content worked before, for some it never worked. I ~~can~~ can't say why and when though.

**Comment by nuclearsandwich** [comment]
> I can say why and when though.

Assuming you mean *can't* here rather than *can*.

The explanation I've proposed for this instance is that it's parsing the logs to read an absolute file path that is only valid inside a container but looking for that path on the host filesystem and rightly not finding it.

If that passes scrutiny from other folks I think we can close this as a wontfix/untargeted for the moment as it is not something we're likely to look into in the determinable future and the workaround is to get the filename and line number from the report table (reading the relative path from the error page if needed) and access the file locally or on GitHub.

**Comment by dirk-thomas** [comment]
This might be related to ros2/ros2#241. Imo it would be a good thing to fix therefore I don't think the ticket should be closed.

**Comment by dirk-thomas** [comment]
Assuming this was resolved by ros2/ci#248 I will go ahead and close this.
